_posts/2019-12-23-Presto-Experiment-with-Graivton-Processor.md

```diff
@@ -0,0 +1,162 @@
+
+```
+# Install Docker
+$ sudo yum update -y
+$ sudo amazon-linux-extras install docker -y
+$ sudo service docker start
+$ sudo usermod -a -G docker ec2-user
+
+# Install docker-compose
+$ sudo yum install python2-pip gcc libffi-devel openssl-devel -y
+$ sudo pip install -U docker-compose
+```
+
+# Performance Comparison
+
+Let's briefly take a look into how the performance provided by the Graviton processor looks like. We are going to use [a1.4xlarge](https://aws.amazon.com/ec2/instance-types/a1/) as a benchmark instance of Graviton processor.
+
+Here is our specification of the benchmark conditions.
+
+- We use the commit [`b0c07249de5c70a70b3037875df4fd0477dec9fc`](https://github.com/prestosql/presto/commit/b0c07249de5c70a70b3037875df4fd0477dec9fc) + the patch previously described.
+- 1 coordinator + 2 worker processes run by [docker-compose](https://docs.docker.com/compose/) on a single instance.
+- We use a1.4xlarge and c5.4xlarge, whose CPU core and memory are the same as a1.4xlarge. And we also compared with m5.2xlarge, whose on-demand instance cost is close to a1.4xlarge.
+- We use [q01, q10, q18, and q20](https://github.com/prestosql/presto/tree/master/presto-benchto-benchmarks/src/main/resources/sql/presto/tpch) run on the TPCH connector. Since the Presto TPCH connector does not access external storage, we can measure pure CPU performance without worrying about network variance.
+- We choose `tiny` and `sf1` as the scaling factor of TPCH connector
+- Our experiment measures the average time of 5 query runtime after 5 times warmup for every query.
+
+#### OpenJDK 8
+Here is the result of our experiment. The vertical axis represents the running time in milliseconds.
+
+![OpenJDK 8 Performance](/assets/blog/presto-experiment-with-graviton-processor/openjdk8-performance.png)
+
+It shows c5.4xlarge achieves the best performance consistently in every case. Compared with m5.2xlarge, the result was switched by the query type. a1.4xlarge and m5.2xlarge are probably competing with each other.
+
+Although we use OpenJDK 8 for this case, it might not be able to generate the code fully optimized for Arm architecture. In general, the later versions, such as [OpenJDK 9 or 11, give us better performance](https://medium.com/@carlosedp/java-benchmarks-on-arm64-17edd8b9ff79).
+
+#### OpenJDK 11
+Let's try to run Presto with OpenJDK 11 again.  There is one thing to do. From JDK 9, the [Attach API](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8180425) was disabled as default. We have found that we needed to allow the usage of attach API by adding the following option in `jvm.config` file, otherwise we will see an error message at the bootstrap phase.
```

Comment:
`-Djdk.attach.allowAttachSelf=true` is a recommended option --
https://github.com/prestosql/presto/blob/8c9b93337d2338f630736165e5fb442fdef2ce49/presto-docs/src/main/sphinx/installation/deployment.rst#L85